libdnf/repo/Repo.cpp

```diff
@@ -0,0 +1,1706 @@
+}
+
+int Repo::Impl::mirrorFailureCB(void * data, const char * msg, const char * url, const char * metadata)
+{
+    if (!data)
+        return 0;
+    auto cbObject = static_cast<RepoCB *>(data);
+    return cbObject->handleMirrorFailure(msg, url, metadata);
+/*    std::cout << "HMF msg: " << msg << "url: " << url << "metadata: "<< metadata << std::endl;
+    return static_cast<int>(LR_CB_OK);*/
+    //LR_CB_ERROR
+};
+
+
+/**
+* @brief Converts the given input string to a URL encoded string
+*
+* All input characters that are not a-z, A-Z, 0-9, '-', '.', '_' or '~' are converted
+* to their "URL escaped" version (%NN where NN is a two-digit hexadecimal number).
+*
+* @param src String to encode
+* @return URL encoded string
+*/
+static std::string urlEncode(const std::string & src)
+{
+    auto noEncode = [](char ch)
+    {
+        return isalnum(ch) || ch=='-' || ch == '.' || ch == '_' || ch == '~';
+    };
+
+    // compute length of encoded string
+    auto len = src.length();
+    for (auto ch : src) {
+        if (!noEncode(ch))
+            len += 2;
+    }
+
+    // encode the input string
+    std::string encoded;
+    encoded.reserve(len);
+    for (auto ch : src) {
+        if (noEncode(ch))
+            encoded.push_back(ch);
+        else {
+            encoded.push_back('%');
+            unsigned char hex;
+            hex = static_cast<unsigned char>(ch) >> 4;
+            hex += hex <= 9 ? '0' : 'a' - 10;
+            encoded.push_back(hex);
+            hex = static_cast<unsigned char>(ch) & 0x0F;
+            hex += hex <= 9 ? '0' : 'a' - 10;
+            encoded.push_back(hex);
+        }
+    }
+
+    return encoded;
+}
+
+/**
+* @brief Format user password string
+*
+* Returns user and password in user:password form. If quote is True,
+* special characters in user and password are URL encoded.
+*
+* @param user Username
+* @param passwd Password
+* @param encode If quote is True, special characters in user and password are URL encoded.
+* @return User and password in user:password form
+*/
+static std::string formatUserPassString(const std::string & user, const std::string & passwd, bool encode)
+{
+    if (encode)
+        return urlEncode(user) + ":" + urlEncode(passwd);
+    else
+        return user + ":" + passwd;
+}
+
+Repo::Impl::Impl(const std::string & id, std::unique_ptr<ConfigRepo> && conf)
+: id(id), conf(std::move(conf)), libsolvRepo(nullptr), timestamp(-1)
+, syncStrategy(SyncStrategy::TRY_CACHE), expired(false) {}
+
+Repo::Impl::~Impl()
+{
+    g_strfreev(mirrors);
+    if (libsolvRepo)
+        libsolvRepo->appdata = nullptr;
+}
+
+Repo::Repo(const std::string & id, std::unique_ptr<ConfigRepo> && conf)
+: pImpl(new Impl(id, std::move(conf))) {}
+
+Repo::~Repo() = default;
+
+void Repo::setCallbacks(std::unique_ptr<RepoCB> && callbacks)
+{
+    pImpl->callbacks = std::move(callbacks);
+}
+
+int Repo::verifyId(const std::string & id)
```

Comment:
What about to verify also an id in Repo constructor? I guess that if incorrect id makes a problem for libsolv or more probably for librepo.